firedrake/plot.py

```diff
@@ -1,373 +1,280 @@
+        for setter, idx in zip(setters, range(coords.shape[1])):
+            try:
+                setter = getattr(axes, setter)
+            except AttributeError:
+                continue
+            amin = coords[:, idx].min()
+            amax = coords[:, idx].max()
+            extra = (amax - amin) / 20
+            if extra == 0.0:
+                # 1D interval
+                extra = 0.5
+            amin -= extra
+            amax += extra
+            setter(amin, amax)
+
+
+def _get_collection_types(gdim, tdim):
+    if gdim == 2:
+        if tdim == 1:
+            # Probably a CircleCollection?
+            raise NotImplementedError("Didn't get to this yet...")
+        elif tdim == 2:
+            return LineCollection, PolyCollection
+    elif gdim == 3:
+        if tdim == 1:
+            raise NotImplementedError("Didn't get to this one yet...")
+        elif tdim == 2:
+            return Line3DCollection, Poly3DCollection
+        elif tdim == 3:
+            return Poly3DCollection, Poly3DCollection
+
+    raise ValueError("Geometric dimension must be either 2 or 3!")
+
+
+def triplot(mesh, axes=None, interior_kw={}, boundary_kw={}):
+    r"""Plot a mesh with a different color for each boundary segment
+
+    :arg mesh: mesh to be plotted
+    :arg axes: axis object on which to plot mesh
+    :arg interior_kw: keyword arguments to matplotlib collection for mesh interior
+    :arg boundary_kw: keyword arguments to matplotlib collection for mesh boundary
     """
-    from matplotlib import pyplot as plt
     gdim = mesh.geometric_dimension()
     tdim = mesh.topological_dimension()
-    if surface:
-        tdim -= 1
-    if tdim not in [1, 2]:
-        raise NotImplementedError("Not implemented except for %d-dimensional meshes", tdim)
-    if gdim == 3:
-        from mpl_toolkits.mplot3d import Axes3D  # noqa: F401
-        from mpl_toolkits.mplot3d.art3d import Line3DCollection as Lines
-        projection = "3d"
-    else:
-        from matplotlib.collections import LineCollection as Lines
-        from matplotlib.collections import CircleCollection as Circles
-        projection = None
+    BoundaryCollection, InteriorCollection = _get_collection_types(gdim, tdim)
+    quad = mesh.ufl_cell().cellname() == "quadrilateral"
+
+    if axes is None:
+        figure = plt.figure()
+        if gdim == 3:
+            axes = figure.add_subplot(projection='3d')
+        else:
+            axes = figure.add_subplot()
+
     coordinates = mesh.coordinates
-    ele = coordinates.function_space().ufl_element()
-    if ele.degree() != 1:
+    element = coordinates.function_space().ufl_element()
+    if element.degree() != 1:
         # Interpolate to piecewise linear.
-        from firedrake import VectorFunctionSpace, interpolate
-        V = VectorFunctionSpace(mesh, ele.family(), 1)
+        V = VectorFunctionSpace(mesh, element.family(), 1)
         coordinates = interpolate(coordinates, V)
```

Comment:
I'm not sure I understand your example -- since f is zero-initialized the mesh will have no interior. I don't know what the right way to plot that would be. This code was to handle cases where the mesh has degree 2 or more. Plotting curved meshes is for a later PR ;)

Reply:
In the example I was trying to find a minimal version of something like changing mesh coordinates as detailed [here](https://www.firedrakeproject.org/mesh-coordinates.html) with the aim of working out what the point of the highlighted line is. But nothing in this PR is broken - plotting the mesh in the existing API also doesn't gives me an empty box (with a dot in the centre) too so you've at least maintained existing behaviour! Marking as resolved.